module/zfs/zfs_vnops.c

```diff
@@ -979,13 +981,497 @@ zfs_get_done(zgd_t *zgd, int error)
+		    &clear_setid_bits_txg, tx);
+
+		zfs_tstamp_update_setup(dstzp, CONTENT_MODIFIED, mtime, ctime);
+
+		/*
+		 * Update the file size (zp_size) if it has changed;
+		 * account for possible concurrent updates.
+		 */
+		while ((dstsize = dstzp->z_size) < dstoffset + size) {
+			(void) atomic_cas_64(&dstzp->z_size, dstsize,
+			    dstoffset + size);
+		}
+
+		error = sa_bulk_update(dstzp->z_sa_hdl, bulk, count, tx);
+
+		zfs_log_clone(zilog, tx, TX_CLONE, dstzp, dstioflag, dstoffset,
+		    size, srcblksz, bps, nbps);
+		dmu_tx_commit(tx);
+
+		kmem_free(bps, sizeof (bps[0]) * nbps);
+
+		if (error != 0)
+			break;
+
+		srcoffset += size;
+		dstoffset += size;
+		length -= size;
+		*donep += size;
+	}
+
+	zfs_znode_update_vfs(dstzp);
+	zfs_rangelock_exit(dstlr);
+	zfs_rangelock_exit(srclr);
+
+	/*
+	 * If we made no progress, return error.
+	 * Otherwise, it's at least a partial write, so it's successful.
+	 */
+	if (*donep == 0) {
+		zfs_exit_two(srczfsvfs, dstzfsvfs);
+		return (error);
+	}
+
+	if (dstioflag & (O_SYNC | O_DSYNC) ||
+	    dstzfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS) {
+		zil_commit(zilog, dstzp->z_id);
+	}
+
+	/*
+	 * TODO: We don't log access time update to ZIL of the source
+	 *       file system.
+	 */
+	ZFS_ACCESSTIME_STAMP(srczfsvfs, srczp);
+	zfs_exit_two(srczfsvfs, dstzfsvfs);
+
+	return (0);
+}
+
+int
+zfs_clone_range_replay(znode_t *zp, uint64_t offset, uint64_t length,
+    uint64_t blksz, const blkptr_t *bps, size_t nbps)
+{
+	zfsvfs_t	*zfsvfs;
+	dmu_buf_impl_t	*db;
+	dmu_tx_t	*tx;
+	int		error;
+	int		count = 0;
+	sa_bulk_attr_t	bulk[4];
+	uint64_t	mtime[2], ctime[2];
+
+	ASSERT3U(offset, <, MAXOFFSET_T);
+	ASSERT3U(length, >, 0);
+	ASSERT3U(nbps, >, 0);
+
+	zfsvfs = ZTOZSB(zp);
+
+	ZFS_ENTER(zfsvfs);
+	ZFS_VERIFY_ZP(zp);
+
+	ASSERT(zfsvfs->z_replay);
+	ASSERT(!zfs_is_readonly(zfsvfs));
+
+	if (!spa_feature_is_enabled(dmu_objset_spa(zfsvfs->z_os),
+	    SPA_FEATURE_BLOCK_CLONING)) {
+		ZFS_EXIT(zfsvfs);
+		return (SET_ERROR(EOPNOTSUPP));
+	}
+
+	if ((offset % blksz) != 0) {
+		ZFS_EXIT(zfsvfs);
+		return (SET_ERROR(EINVAL));
+	}
+
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_MTIME(zfsvfs), NULL, &mtime, 16);
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_CTIME(zfsvfs), NULL, &ctime, 16);
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_SIZE(zfsvfs), NULL,
+	    &zp->z_size, 8);
+	SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_FLAGS(zfsvfs), NULL,
+	    &zp->z_pflags, 8);
+
+	/*
+	 * Start a transaction.
+	 */
+	tx = dmu_tx_create(zfsvfs->z_os);
+
+	dmu_tx_hold_sa(tx, zp->z_sa_hdl, B_FALSE);
+	db = (dmu_buf_impl_t *)sa_get_db(zp->z_sa_hdl);
+	DB_DNODE_ENTER(db);
+	dmu_tx_hold_write_by_dnode(tx, DB_DNODE(db), offset, length);
+	DB_DNODE_EXIT(db);
+	zfs_sa_upgrade_txholds(tx, zp);
+	error = dmu_tx_assign(tx, TXG_WAIT);
+	if (error != 0) {
+		dmu_tx_abort(tx);
+		ZFS_EXIT(zfsvfs);
+		return (error);
+	}
+
+	if (zp->z_blksz < blksz)
+		zfs_grow_blocksize(zp, blksz, tx);
+
+	dmu_brt_addref(zfsvfs->z_os, zp->z_id, offset, length, tx, bps, nbps);
+
```

Comment:
You need to call zfs_log_clone somewhere here,
so that zil_replaying() gets called.
That function is actually side-effectful:
it is used to update the ZIL header to record
replay progress.

To not forget that, the usual recipe is to have just
one zfs_OP function that calls zfs_log_OP.
The replay callback then just calls zfs_OP,
which will call zfs_log_OP again, but
zfs_log_OP early-returns if zil_replaying() is true.

If there's nothing stopping you from just having one function, I suggest you stick to the pattern.
If there is, explain why in a function-level comment at the top.

Reply:
Unfortunately I cannot have one function. The regular clone function operates on two objects: source and destination. When we replaying we only have the destination object and BPs from the source object, so the implementations are very different. I've added zil_replaying() though.